enhancements/cluster-logging/content-filter.md

```diff
@@ -0,0 +1,261 @@
+  type: prune
+  prune:
+    remove:
+	  - .kubernetes.flat_labels
+	  - .kubernetes.pod_labels
+	  - .kubernetes.namespace_labels
+	  - .kubernetes.annotations
+
+```
+
+
+#### Drop filters
+
+##### API
+
+``` yaml
+spec:
+  filters:
+  - name:                #user-selected name
+    type: drop
+    drop:
+	  rules:
+		- path:      # JSON path to the field
+		  action:    # enum: keep|drop
+		  # Following are alternatives, at most one:
+		  in:         # array, true if field value == any element
+		  notIn:      # array, true if field value != all elements
+		  matches:    # regular expression match
+		  notMatches: # regular expression does not match
+		  equals:     # equality - any type
+		  notEquals:  # inequality - any type
+		  less:       # numeric less-than
+		  greater:    # numeric greater-than
+```
+
+Note:
+- Rules are tried in order until a rule matches. Remaining rules are ignored.
+- The action of the matching rule is the outcome of the filter.
+- If there is an error evaluating a rule (missing field, mismatched type...), that is equivalent to not matching. \
+  Continue to next rule.
+- If no rule matches the default outcome is `keep`.
+- The default outcome can be changed to `drop` by adding a final unconditional rule `{ action: drop }`
+
+##### Boolean completeness
+
+Consider rules as propositions. Each test has a negative, so we can negate these propositions.
+Any boolean expression can be reduced to an AND or OR of individual propositions or their negation.
+
+A sequence of "keep" followed by an unconditional "drop" is an OR clause, keep if and only if some rule matchers.
+A sequence of "drop" is a NAND clause, keep if an only if no rule matches.
+
+Thus this mechanism is boolean complete.
+
+##### Example
+
+``` yaml
+filters:
+  - name: important
+    type: drop
+    drop: # Type of filter, selets records to keep/drop
+      rules:
+      - field: .kubernetes.namespace_name
+        matches: so-important # Keep everything from this namespace regardless of level.
+        action: keep
+      - field: .level # Keep important levels
+        matches: warn(ing)?|err(or)?|critical
+        action: keep
+      - action: drop # Drop everything else.
+
+  - name: slim
+    prune: # Type of filter, prunes unwanted data
+      removeOthers: [.kubernetes, .message, .structured] # Remove all other fields.
+      remove: [.kubernetes.annotations] # Remove this sub-field.
+```
+
+##### Vector configuration
+
+See the implementation of the "kubeAPIAUdit" filter type, just the "rules" section.
+The tests are different but the iteration over a list of rules with actions is very similar.
+
+##### Pre-defined filters
+
+Just like default inputs ("application", "audit", "infrastructure") and outputs ("default"), we can define reserved names for pre-defined filters.
+These are equivalent to user-defined filters but pre-defined for convenience.
+
+- "DropDebug": `rules: { field: .level , in: [trace, debug] , action: drop }`
```

Comment:
@alanconway are trace, debug, info case sensitive? I suppose they are not, then,  if they are not, should TRACE, DEBUG, INFO being added?

Reply:
The Viaq 'level' field is derived from the original log line by a bunch of regexp matching heuristics that allow a bunch of different caps/spelling/abbreviation variations; the 'level' field itself is (supposed to be) normalized to lower case syslog RFC keywords.